virtual_rainforest/soil/model.py

```diff
@@ -40,57 +36,67 @@ class SoilModel(BaseModel, model_name="soil"):
 
         if no_layers < 1:
-            LOGGER.error("There has to be at least one soil layer in the soil model!")
+            log_and_raise(
+                "There has to be at least one soil layer in the soil model!",
+                InitialisationError,
+            )
         elif no_layers != int(no_layers):
-            LOGGER.error("The number of soil layers must be an integer!")
+            log_and_raise(
+                "The number of soil layers must be an integer!", InitialisationError
+            )
 
         super(SoilModel, self).__init__(update_interval, **kwargs)
         self.no_layers = int(no_layers)
         # Save variables names to be used by the __repr__
         self._repr.append("no_layers")
 
     @classmethod
-    def factory(cls, config: dict[str, Any]) -> SoilModel:
+    def factory(cls, config: dict[str, Any]) -> Optional[SoilModel]:
         """Factory function to initialise the soil model.
 
         This function unpacks the relevant information from the configuration file, and
-        then uses it to initialise the model.
+        then uses it to initialise the model. If any information from the config is
+        invalid rather than returning an initialised model instance None is returned.
 
         Args:
             config: The complete (and validated) virtual rainforest configuration.
 
-        Raises:
-            InitialisationError: If the information required to initialise the model
-                either isn't found, or isn't of the correct type.
         """
+
+        # Assume input is valid until we learn otherwise
+        valid_input = True
         try:
-            raw_interval = Quantity(config["core"]["timing"]["min_time_step"]).to(
+            raw_interval = pint.Quantity(config["core"]["timing"]["min_time_step"]).to(
                 "minutes"
             )
             # Round raw time interval to nearest minute
             update_interval = timedelta64(int(raw_interval.magnitude), "m")
             no_layers = config["soil"]["no_layers"]
         except KeyError as e:
-            log_and_raise(
+            valid_input = False
+            LOGGER.error(
                 f"Configuration is missing information required to initialise the soil "
-                f"model. The first missing key is {str(e)}.",
-                InitialisationError,
+                f"model. The first missing key is {str(e)}."
             )
-        except ValueError as e:
-            log_and_raise(
+        except (ValueError, pint.errors.DimensionalityError) as e:
+            valid_input = False
+            LOGGER.error(
                 f"Configuration types appear not to have been properly validated. This "
                 f"problem prevents initialisation of the soil model. The first instance"
-                f" of this problem is as follows: {str(e)}",
-                InitialisationError,
+                f" of this problem is as follows: {str(e)}"
             )
```

Comment:
To me, this check belongs in the `__init__` method. If someone calls `SoilModel('23 martian days)'` programatically, then that has to throw a `DimensionalityError`. I think this method should only handle unpacking `config` and feeding it into a class instance. So `KeyErrors` but not stuff which needs validating in `__init__`.

Reply:
So the `__init__` would take in e.g. `"23 martian days"` as a string, and then the conversion to a `Quantity` (and potential throwing of a `DimensionalityError`) would take place there?

Reply:
Oh - huh. Hadn't twigged that SoilModel is expecting an `np.timedelta64` and so this is specifically converting from config user friendliness to programming API.

Reply:
Shouldn't the missing keys stuff be handled by the schema validation though?

Reply:
Yeah the key stuff should be handled by the schema validation when the whole process is run. I put in additional checking here to cover for cases where the user inadvertently modified the configuration, but I'm not 100% sure whether or not that additional checking is sensible

Reply:
Do you reckon I should remove the missing key checking? And would you recommend moving the `np.timedelta64` conversion to the `__init__`?